src/inmanta/ast/statements/__init__.py

```diff
@@ -149,15 +234,103 @@ def get_dataflow_node(self, graph: DataflowGraph) -> dataflow.NodeReference:
 
 
 class Resumer(ExpressionStatement):
+    """
+    Resume on a set of requirement variables' values when they become ready (i.e. they are complete).
+    """
+
     def resume(self, requires: Dict[object, object], resolver: Resolver, queue: QueueScheduler, target: ResultVariable) -> None:
         pass
 
 
 class RawResumer(ExpressionStatement):
+    """
+    Resume on a set of requirement variables when they become ready (i.e. they are complete).
+    """
+
     def resume(self, requires: Dict[object, ResultVariable], resolver: Resolver, queue: QueueScheduler) -> None:
         pass
 
 
+class VariableReferenceHook(RawResumer):
+    """
+    Generic helper class for adding a hook to a variable (ResultVariable) object. Supports both plain variables and instance
+    attributes. Calls variable resumer with the variable object as soon as it's available.
+    """
+
+    def __init__(
+        self,
+        instance: Optional["Reference"],
+        name: str,
+        variable_resumer: "VariableResumer",
+    ) -> None:
+        super().__init__()
+        self.instance: Optional["Reference"] = instance
+        self.name: str = name
+        self.variable_resumer: "VariableResumer" = variable_resumer
```

Comment:
Why this Resumer in a Resumer structure? 

In my mind, a resumer is a kind of synthetic statement that introduces a temporary result variable to `break` a large expression into sub-expression that can be scheduled independently. The  temporary result variable acts a s a Future that ties the producer and consumer together.  This introduces a different mode of chaining resumers directly?

Reply:
If I understand your question correctly you're asking why I refactored this from an ABC class with a couple of subclasses to a concrete resumer class with an ABC `VariableResumer`? The main reason is to make abstraction of whatever is done with the `ResultVariable` object once it's obtained. I found it made more sense to split this conceptually in a waiter for the resultvariable and an action on that resultvariable, than the mix that was currently in place, which involved two execution phases for the same object (first wait for the instance, then cache the variable and continue execution for the variable).

In my mind this composition makes for a clear separation of responsibilities: the one object waits for an instance and fetches its attribute, then delegates to the other object that knows what to do with that variable.

Does my explanation make sense? Do you think I should revert this?

Side note: the `VariableResumer` is a bit different from the other `Resumer` classes (I didn't even make it inherit from them): it doesn't take requires, just a resultvariable. I started out with "action" rather than resumer but I didn't like it as much. Perhaps yet another name could be useful? It's just a class that works on or with a resultvariable in some way.

Reply:
Other side note: my idea was that attribute assignment could be refactored at some point to use these same abstractions, because conceptually it just does the same: wait for the instance, fetch the attribute, assign to it.